sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -858,6 +859,30 @@ object PushPredicateThroughJoin extends Rule[LogicalPlan] with PredicateHelper {
 }
 
 /**
+  * Push [[Limit]] operators through [[Join]] operators, iff the join type is outer joins.
+  * Adding extra [[Limit]] operators on top of the outer-side child/children.
+  */
+object PushLimitThroughOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
```

Comment:
Need to refine the idea. 

For the `full outer` join, 
1) if both sides have `limit`. Do nothing. 
2) if only one side has a `limit`, add extra `limit` to that side. 
3) if both sides do not have a limit,  add extra `limit` to one side which has a higher `statistics`

Is it better?